client/lib.go

```diff
@@ -30,18 +30,82 @@ const (
 	contentTypeJSON = "application/json"
 )
 
-// NewGoShimmerAPI returns a new *GoShimmerAPI with the given baseURL and httpClient.
-func NewGoShimmerAPI(baseURL string, httpClient ...http.Client) *GoShimmerAPI {
-	if len(httpClient) > 0 {
-		return &GoShimmerAPI{baseURL: baseURL, httpClient: httpClient[0]}
+// Options define the options of a GoShimmer API client.
+type Options struct {
+	// The initial committee of the DRNG.
+	BasicAuth *BasicAuth
+	// The initial randomness of the DRNG.
+	HTTPClient *http.Client
+}
+
+// Option is a function which sets the given option.
+type Option func(*Options)
+
+// BasicAuth defines the basic-auth struct.
+type BasicAuth struct {
+	Enabled  bool   `json:"enabled,omitempty"`
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+}
+
+// NewBasicAuth returns a new BasicAuth.
+func NewBasicAuth(username, password string) *BasicAuth {
+	return &BasicAuth{
+		Enabled:  true,
+		Username: username,
+		Password: password,
+	}
+}
+
+// IsEnabled returns the enabled state of a given BasicAuth.
+func (b *BasicAuth) IsEnabled() bool {
+	if b != nil {
+		return b.Enabled
+	}
+	return false
+}
+
+// Credentials returns the username and password of a given BasicAuth.
+func (b *BasicAuth) Credentials() (username, password string) {
+	return b.Username, b.Password
+}
+
+// SetBasicAuth sets the basic-auth.
+func SetBasicAuth(b *BasicAuth) Option {
+	return func(args *Options) {
+		args.BasicAuth = b
+	}
+}
+
+// SetHTTPClient sets the http Client.
+func SetHTTPClient(c *http.Client) Option {
+	return func(args *Options) {
+		args.HTTPClient = c
+	}
+}
+
+// NewGoShimmerAPI returns a new *GoShimmerAPI with the given baseURL and options.
+func NewGoShimmerAPI(baseURL string, setters ...Option) *GoShimmerAPI {
+	args := &Options{}
+
+	for _, setter := range setters {
+		setter(args)
+	}
+	if args.HTTPClient == nil {
+		args.HTTPClient = &http.Client{}
+	}
```

Comment:
Don't need this check if we have `newOptions`.
```
func newOptions() *Options {
    return &Options{
         BasicAuth: &BasicAuth{},
         HTTPClient: &http.Client{},
    }
}
```

Reply:
@jkrvivian good suggestions! I've refactored the way we handle options, so that now feels more intuitive (I hope)!

Reply:
That's even better! :smile: